### PR TITLE
feat: add popper.js to action-popover - FE-3265

### LINKED ISSUE #3321: Last PopoverItem not displayed
PopoverItem not displayed if last 

### Current behaviour
IF we have ActionPopoverItem Items with component in  below like GridContainer with Button for example : 
The latest items aren't displayed correctly .
In the sendBox https://codesandbox.io/s/gallant-bhaskara-c1ujf?file=/src/index.js:2693-2706 
You can see that Test3 is not displayed.
But if we resize the page (only in the sandbox) => We display correctly the all of the items.
We have the same problem with list of entities (test3 is not displayed) : 
image

### Expected behaviour

All ActionPopoverItem  itmes should be displayed 

### Reproducible example

https://codesandbox.io/s/gallant-bhaskara-c1ujf?file=/src/index.js:2693-2706

### Your environment

| Software         | Version(s) |
| ---------------- | ---------- |
| carbon-react     | 44.1.0
| carbon-factory   |
| react-scripts    | 
| React            |
| Browser          |
| npm              |
| Operating System |

### PR DESCRIPTION
**Pull request by dawidzarzycki**
### Proposed behaviour
Add popper.js to action-popover and use it mechanics to show hidden menu. It will fix many issues related to z-index/overflow

Fixes #3321

### Current behaviour
We use our implementation of showing hidden submenu and it doesn't work well

### Checklist

- [x] Commits follow our style guide
- [ ] <del> Screenshots are included in the PR if useful
- [ ] <del> All themes are supported if required
- [x] Unit tests added or updated if required
- [ ] Cypress automation tests added or updated if required
- [ ] <del> Storybook added or updated if required
- [ ] <del> Typescript `d.ts` file added or updated if required
- [ ] <del> Carbon implementation and Design System documentation are congruent

### Testing instructions
We should check all stories in action-popover `storybook/design-system/action-popover`
`with overflow hidden container` shows exactly the same scenario as described in github issue.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -156,6 +157,7 @@
     "react-dnd-html5-backend": "^10.0.2",
     "react-dnd-legacy": "npm:react-dnd@^2.6.0",
     "react-dnd-touch-backend": "0.2.7",
+    "react-popper": "^2.2.4",
```
Comment (dawidzarzycki) [inline package.json:160]: react-popper is no more needed, I will remove it

File: src/components/accordion/accordion.stories.mdx [hunk 2]
```diff
@@ -70,6 +82,80 @@ const MyComponent = () => (
 
 ## Visuals
 
+## FAILS
```
Comment (dawidzarzycki) [inline src/components/accordion/accordion.stories.mdx:85]: this is no more needed story example, I will remove it

File: src/components/action-popover/action-popover.spec.js [hunk 3]
```diff
@@ -41,7 +44,7 @@ describe("ActionPopover", () => {
   const onClose = jest.fn();
   const onClickWrapper = (arg) => () => onClick(arg);
 
-  function render(props = {}, renderer = mount) {
+  async function render(props = {}, renderer = mount) {
```
Comment (dawidzarzycki) [inline src/components/action-popover/action-popover.spec.js:47]: Probably not needed, I will remove it

File: src/components/action-popover/action-popover.spec.js [hunk 4]
```diff
@@ -253,16 +256,69 @@ describe("ActionPopover", () => {
...
-    render({ rightAlignMenu: true });
-    const { menu } = getElements();
-    assertStyleMatch(
-      {
-        right: undefined,
-        left: "0",
-      },
-      menu
-    );
+  describe("it shoud render correct popper props", () => {
```
Comment (DipperTheDan) [inline src/components/action-popover/action-popover.spec.js:259]: Spelling mistake on "shoud". Should be "should".

File: src/components/action-popover/action-popover.spec.js [hunk 5]
```diff
@@ -253,16 +256,69 @@ describe("ActionPopover", () => {
...
+    let menu;
+    let button;
+
+    it("if prop `top` is passed", () => {
```
Comment (DipperTheDan) [inline src/components/action-popover/action-popover.spec.js:263]: You've repeated this block a few times between here and line 308. Could the same outcome be achieved using a `it.each` instead? If you've done it this way for readability purposes, that's fine too.

File: src/components/action-popover/action-popover.component.js [hunk 6]
```diff
@@ -112,32 +130,40 @@ const ActionPopover = ({
...
-      ref={button}
-      {...rest}
-    >
-      {menuButton()}
-      <ActionPopoverMenu
-        data-component="action-popover"
-        role="menu"
-        ref={menu}
-        {...menuProps}
+    <>
```
Comment (mkrds) [inline src/components/action-popover/action-popover.component.js:138]: This <> `React.Fragment` is not needed here

File: src/components/action-popover/action-popover.component.js [hunk 7]
```diff
@@ -112,32 +130,40 @@ const ActionPopover = ({
     placement,
   };
 
+  useEffect(() => {
+    createPopper(button.current, menu.current, { placement: valueMapper() });
+  });
+
```
Comment (mkrds) [inline src/components/action-popover/action-popover.component.js:136]: Read this section https://popper.js.org/docs/v2/tutorial/#performance it looks like Popper should be destroyed after the component unmounts

File: src/components/action-popover/action-popover.component.js [hunk 8]
```diff
@@ -25,6 +28,25 @@ const ActionPopover = ({
   const [guid] = useState(createGuid());
   const button = useRef();
   const menu = useRef();
+  // const elementDOM = useRef(document.createElement("div"));
```
Comment (jamime) [inline src/components/action-popover/action-popover.component.js:31]:
```suggestion
```

File: src/components/action-popover/action-popover.component.js [hunk 9]
```diff
@@ -25,6 +28,25 @@ const ActionPopover = ({
...
+  const popperInstance = useRef();
+  const isMounted = useRef(false);
+
+  const valueMapper = useCallback(() => {
```
Comment (jamime) [inline src/components/action-popover/action-popover.component.js:35]: I think this should be `useMemo` as it accepts no arguments. It's debatable if the overhead of `useMemo` or `useCallback` provides any performance improvement when the logic is simple.
```suggestion
  const mappedPlacement = useMemo(() => {
```

File: src/components/action-popover/action-popover.component.js [hunk 10]
```diff
@@ -115,6 +137,19 @@ const ActionPopover = ({
     placement,
   };
 
+  useEffect(() => {
+    if (!isMounted.current) {
+      popperInstance.current = createPopper(button.current, menu.current, {
+        placement: valueMapper(),
```
Comment (jamime) [inline src/components/action-popover/action-popover.component.js:143]:
```suggestion
        placement: mappedPlacement,
```

File: src/components/action-popover/action-popover.spec.js [hunk 11]
```diff
@@ -396,15 +485,15 @@ describe("ActionPopover", () => {
...
+        const { menubutton } = getElements();
 
         menubutton.simulate("click");
 
         act(() => {
           document.dispatchEvent(
             new CustomEvent("click", {
               detail: {
-                enzymeTestingTarget: items.first().getDOMNode(),
+                enzymeTestingTarget: menubutton.getDOMNode(),
```
Comment (jamime) [inline src/components/action-popover/action-popover.spec.js:496]: Why did we change this test, it previously use to click on the first menu item, but now it's clicking on the `menuButton`. The intention here was to test the event propagation/composedPath.

File: src/components/action-popover/action-popover.stories.mdx [hunk 12]
```diff
@@ -653,6 +654,34 @@ used actions, in this example "manage devices".
   </Story>
 </Preview>
 
+## in overflow hidden container
```
Comment (jamime) [inline src/components/action-popover/action-popover.stories.mdx:657]: We've got some odd `z-index` here 
image

image

File: src/components/action-popover/action-popover.component.js [hunk 13]
```diff
@@ -184,7 +221,7 @@ ActionPopover.propTypes = {
   /** Render a custom menu button to override default ellipsis icon */
   renderButton: PropTypes.func,
   /** Set whether the menu should open above or below the button */
-  placement: PropTypes.oneOf(["bottom", "top"]),
+  placement: PropTypes.oneOf(["top", "bottom"]),
```
Comment (jamime) [inline src/components/action-popover/action-popover.component.js:224]: This was actually introduced for when the popover would be displayed on the last row of a table, and that table was at the bottom of the page. Won't popper automatically cater for this, if so I think we can deprecate this prop.

**Review by jamime — COMMENTED** [review]
I wanted to check, is this the expected behaviour. It seems to re-position to above the table which means all of the menu items are off screen. If it was to remain below the table we would still be able to see the first option. I've tried to show the behaviour in the video below.

https://user-images.githubusercontent.com/2328042/102997190-63298b80-451c-11eb-8df2-64f0bfc2d315.mov

**Comment by DipperTheDan** [comment]
Good day. I've noticed a styling bug. 

This is the style on `master`.
Screenshot 2021-01-07 at 11 33 47

where as on your branch it looks like this:
Screenshot 2021-01-07 at 11 33 33

File: src/components/action-popover/action-popover.spec.js [hunk 14]
```diff
@@ -177,14 +177,13 @@ describe("ActionPopover", () => {
...
-  function createNodeMock() {
-    return {
-      addEventListener: jest.fn(),
-      removeEventListener: jest.fn(),
-      focus: jest.fn(),
-    };
+  function openMenu() {
+    act(() => {
+      wrapper.find(MenuButton).simulate("click");
+      jest.runAllTimers();
```
Comment (jamime) [inline src/components/action-popover/action-popover.spec.js:183]: Hopefully we can remove the timers when we refactor the ref.

**Review by jamime — CHANGES_REQUESTED** [review]

File: src/components/action-popover/action-popover.style.js [hunk 15]
```diff
@@ -78,7 +75,7 @@ const MenuButton = styled.div`
     cursor: pointer;
   }
   width: fit-content;
-  margin: auto;
+  /* margin: auto; */
```
Comment (DipperTheDan) [inline src/components/action-popover/action-popover.style.js:78]: Is it this commented out piece causing the styling bug I commented about here?
https://github.com/Sage/carbon/pull/3508#issuecomment-756064206

File: src/components/action-popover/action-popover-item.component.js [hunk 16]
```diff
@@ -21,201 +20,205 @@ import ActionPopoverContext from "./action-popover-context";
...
+      );
+    }
+  }, [submenu, spacing, placement]);
 
-    useLayoutEffect(() => {
-      alignSubmenu();
-    }, [alignSubmenu]);
+  const setRef = useCallback(
+    (element) => {
+      setTimeout(() => {
```
Comment (jamime) [inline src/components/action-popover/action-popover-item.component.js:65]: We need to clear this timeout if the component is unmouted.

**Review by jamime — CHANGES_REQUESTED** [review]
We're keeping the `setTimeout` for now as it's the only solution that works correctly. We might revisit this in the future if we identify any issues.

**Review by edleeks87 — APPROVED** [review]

**Review by jamime — APPROVED** [review]

**Review by edleeks87 — DISMISSED** [review]

**Review by edleeks87 — DISMISSED** [review]

**Review by jamime — DISMISSED** [review]

**Review by edleeks87 — APPROVED** [review]

**Review by jamime — APPROVED** [review]

**Review by edleeks87 — APPROVED** [review]

**Review by edleeks87 — DISMISSED** [review]

File: cypress/support/step-definitions/action-popover-steps.js [hunk 17]
```diff
@@ -15,7 +15,7 @@ Then('Action Popover element is visible', () => {
 });
 
 Then('Action Popover element is not visible', () => {
-  actionPopover().should('not.be.visible');
```
Comment (KatarzynaQA) [inline cypress/support/step-definitions/action-popover-steps.js:18]: Should be `not.be.visible`, why did you change it ?

File: cypress/features/regression/designSystem/actionPopover.feature [hunk 18]
```diff
@@ -0,0 +1,9 @@
+Feature: Design System Action Popover component
+  I want to test Action Popover component
+
+  @positive
+  Scenario: Invoking Action Popover component in a hidden container
+    Given I open "Design System Action Popover" component page "in_overflow_hidden_container" in no iframe
+    When I expand accordionRow using "Enter" key
```
Comment (KatarzynaQA) [inline cypress/features/regression/designSystem/actionPopover.feature:7]:
```suggestion
     And I expand accordionRow using "Enter" key
```

File: cypress/features/regression/designSystem/actionPopover.feature [hunk 19]
```diff
@@ -0,0 +1,9 @@
...
+      And I click the menu button element in noiFrame
```
Comment (KatarzynaQA) [inline cypress/features/regression/designSystem/actionPopover.feature:8]:
```suggestion
      When I click the menu button element in noiFrame
```

File: cypress/support/step-definitions/action-popover-steps.js (hunk 17 above)
Comment (mkrds) [reply cypress/support/step-definitions/action-popover-steps.js:18]: It has to be `not.exist` 
https://files.slack.com/files-pri/T029WF4K9-F01J38HDHQW/image.png

**Review by edleeks87 — APPROVED** [review]

**Review by jamime — APPROVED** [review]

**Review by KatarzynaQA — APPROVED** [review]
